feat(avm/app/managed-environment): add certificate key vault properties - `avm/res/app/managed-environment`

robvangeloven opened the pull request:
## Description

<!--
>Thank you for your contribution !
> Please include a summary of the change and which issue is fixed.
> Please also include the context.
> List any dependencies that are required for this change.

Fixes #123
Fixes #456
Closes #123
Closes #456
-->

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
|   [![avm.res.app.managed-environment](https://github.com/robvangeloven/bicep-registry-modules/actions/workflows/avm.res.app.managed-environment.yml/badge.svg?branch=feature%2F2642-add-certificate-key-vault-properties)](https://github.com/robvangeloven/bicep-registry-modules/actions/workflows/avm.res.app.managed-environment.yml)       |

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [X] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [X] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change
- [X] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

hundredacres commented:
@robvangeloven Can you add a link to your pipeline?

hundredacres commented:
@robvangeloven You should either update the Max test to include testing this change or create a new one. Let me know if you need help with that.

robvangeloven commented:
> @robvangeloven Can you add a link to your pipeline?

Link here: https://github.com/robvangeloven/bicep-registry-modules/actions/runs/9952753081, though my pipeline throws an error. Looking into that

robvangeloven commented:
> > @robvangeloven Can you add a link to your pipeline?
> 
> Link here: https://github.com/robvangeloven/bicep-registry-modules/actions/runs/9952753081, though my pipeline throws an error. Looking into that

Issue was easy enough, not enough rights for the service principal

robvangeloven commented:
> @robvangeloven You should either update the Max test to include testing this change or create a new one. Let me know if you need help with that.

Yea, I'm going to need some help with that one. I see there's no test for dnsSuffix either so I don't have anything to build upon

AlexanderSehr reviewed (commented):
Hey @robvangeloven,
2 suggestions from my side:
- For all that is holy, please run the [.Platform - Toggle AVM workflows](https://github.com/robvangeloven/bicep-registry-modules/actions/workflows/platform.toggle-avm-workflows.yml) workflow in your fork at least once to disable all workflows and avoid them being triggered upon each change in main. You can enable workflows on demand 💪 
- To the PR at hand, as per other comments, it would be great to get a test in as well, if possible

hundredacres commented:
@robvangeloven I swear I saw this option in the docs last week, but now I'm not seeing it there: https://learn.microsoft.com/en-us/azure/templates/microsoft.app/managedenvironments?pivots=deployment-language-bicep#customdomainconfiguration

Can you confirm if this is an available option still?

hundredacres commented:
> @robvangeloven I swear I saw this option in the docs last week, but now I'm not seeing it there: https://learn.microsoft.com/en-us/azure/templates/microsoft.app/managedenvironments?pivots=deployment-language-bicep#customdomainconfiguration
> 
> Can you confirm if this is an available option still?

I was looking at documentation for the wrong API version.

robvangeloven commented:
> Hey @robvangeloven, 2 suggestions from my side:
> 
> * For all that is holy, please run the [.Platform - Toggle AVM workflows](https://github.com/robvangeloven/bicep-registry-modules/actions/workflows/platform.toggle-avm-workflows.yml) workflow in your fork at least once to disable all workflows and avoid them being triggered upon each change in main. You can enable workflows on demand 💪
> * To the PR at hand, as per other comments, it would be great to get a test in as well, if possible

Sorry for the radio silence, just got back. Will pick it up asap. And yes, I shall run that script. My subscription was suddenly without credits :-P

hundredacres commented:
@robvangeloven I added tests that would support your addition in my own branch if you wanted to take a look and pull them into yours: https://github.com/hundredacres/bicep-registry-modules/tree/feat/issue/2719

robvangeloven commented:
> @robvangeloven I added tests that would support your addition in my own branch if you wanted to take a look and pull them into yours: https://github.com/hundredacres/bicep-registry-modules/tree/feat/issue/2719

Will take a look today! Sorry for the delay, it was suddenly crunch time and had zero spare time

robvangeloven commented:
> @robvangeloven I added tests that would support your addition in my own branch if you wanted to take a look and pull them into yours: https://github.com/hundredacres/bicep-registry-modules/tree/feat/issue/2719

Added the tests in my branch :-)

hundredacres commented:
> > @robvangeloven I added tests that would support your addition in my own branch if you wanted to take a look and pull them into yours: https://github.com/hundredacres/bicep-registry-modules/tree/feat/issue/2719
> 
> Added the tests in my branch :-)

@robvangeloven Get a new pipeline run kicked off with the latest changes so we can confirm a successful deploy.

robvangeloven commented:
> > > @robvangeloven I added tests that would support your addition in my own branch if you wanted to take a look and pull them into yours: https://github.com/hundredacres/bicep-registry-modules/tree/feat/issue/2719
> > 
> > 
> > Added the tests in my branch :-)
> 
> @robvangeloven Get a new pipeline run kicked off with the latest changes so we can confirm a successful deploy.

Getting errors I can't place:

FailedIdentityOperation

Identity operation for resource '/subscriptions/resourceGroups/dep-avm-app.managedenvironments-amemax-rg/providers/Microsoft.App/managedEnvironments/avmamemax001' failed with error 'Failed to perform resource identity operation. Status: 'BadRequest'. Response: '{\"error\":{\"code\":\"BadRequest\",\"message\":\"\"}}'.'.

pipeline link: https://github.com/robvangeloven/bicep-registry-modules/actions/runs/10620968388

ReneHezser commented:
@robvangeloven 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

robvangeloven commented:
> @robvangeloven We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please update your branch and run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.
> 
> Please reach out if any support is needed.

@ReneHezser 

I thought I ran it, but I'll redo the steps you linked.

I'm still stuck on the issue I mentioned before though: I'm getting bad request errors when running the tests and I have no idea where they're coming from.

ReneHezser commented:
> @ReneHezser
> 
> I thought I ran it, but I'll redo the steps you linked.

Please also update your branch. It is way behind and won't update to the newest Readme schema.

eriqua commented:
> > @ReneHezser
> > I thought I ran it, but I'll redo the steps you linked.
> 
> Please also update your branch. It is way behind and won't update to the newest Readme schema.

Hey @robvangeloven just echoing what Rene said above. Please consider pulling latest from upstream, ensure your environment is cleaned up and try rerunning the pipeline again. I see from your badge that the latest run is from 2 months ago, worth a retry 🙂

robvangeloven commented:
> > > @ReneHezser
> > > I thought I ran it, but I'll redo the steps you linked.
> > 
> > 
> > Please also update your branch. It is way behind and won't update to the newest Readme schema.
> 
> Hey @robvangeloven just echoing what Rene said above. Please consider pulling latest from upstream, ensure your environment is cleaned up and try rerunning the pipeline again. I see from your badge that the latest run is from 2 months ago, worth a retry 🙂

Hi @ReneHezser ,

Took a bit longer then planned, but everything is again where it should be. Still getting the bad request operation and don't really have an idea where to look:
```json
{
    "code": "DeploymentFailed",
    "target": "/subscriptions/xxxxxxxxxx/resourceGroups/dep-avm-app.managedenvironments-amemax-rg/providers/Microsoft.Resources/deployments/df555otrxohxs-test-amemax-init",
    "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.",
    "details": [{
            "code": "FailedIdentityOperation",
            "message": "Identity operation for resource '/subscriptions/xxxxxxxx/resourceGroups/dep-avm-app.managedenvironments-amemax-rg/providers/Microsoft.App/managedEnvironments/avmamemax001' failed with error 'Failed to perform resource identity operation. Status: 'BadRequest'. Response: '{\"error\":{\"code\":\"BadRequest\",\"message\":\"\"}}'.'."
        }
    ]
}

```

AlexanderSehr commented on avm/res/app/managed-environment/main.bicep:
```diff
@@ -321,8 +321,22 @@ type roleAssignmentType = {
 
 type certificateKeyVaultPropertiesType = {
   @description('Required. The resoource ID of the identity. This is the identity that will be used to access the key vault.')
```
```suggestion
  @description('Required. The resource ID of the identity. This is the identity that will be used to access the key vault.')
```

AlexanderSehr commented on avm/res/app/managed-environment/main.bicep:
```diff
@@ -321,8 +321,22 @@ type roleAssignmentType = {
 type certificateKeyVaultPropertiesType = {
   @description('Required. The resoource ID of the identity. This is the identity that will be used to access the key vault.')
-  identityResourceId: string
+  identity: string
```
Hm ... as per the specs, this should actually say `resourceId` as it did before. In these cases, where the RP expects a different name, we'd usually map the value in the `resource` block. I.e.,:
```bicep
customDomainConfiguration: {
   certificatePassword: certificatePassword
   certificateValue: !empty(certificateValue) ? certificateValue : null
   dnsSuffix: dnsSuffix
   certificateKeyVaultProperties: {
      identity: certificateKeyVaultProperties.identityResourceId
      keyVaultUrl: certificateKeyVaultProperties.keyVaultUrl
   }
}
```

AlexanderSehr commented on avm/res/app/managed-environment/tests/e2e/max/main.test.bicep:
```diff
@@ -66,6 +69,11 @@ module testDeployment '../../../main.bicep' = [
+      dnsSuffix: 'contoso.com'
+      certificateKeyVaultProperties: {
+        identity: nestedDependencies.outputs.managedIdentityResourceId
+        keyVaultUrl: '${nestedDependencies.outputs.keyVaultUri}secrets/${split(nestedDependencies.outputs.certificateSecretUrl, '/')[4]}'
```
By checking the deployment in Azure, can you confirm if the expected value is passed into the resource-group-level deployment called `.*-test-amemax-.*` nested in the subscription-level deployment `a-r-a-me-max-t1-.*` for the parameter `keyVaultUrl`

robvangeloven replied:
I seem to have fixed the test run issue by fixing the naming issue.  However running the tests concurrently poses issues, running them one works

robvangeloven replied:
False negative: tests run just fine. Must've been a hickup in my environment
